apisix/upstream.lua

```diff
@@ -115,8 +122,11 @@ local function create_checker(upstream)
     end
 
     if healthcheck_parent.checker then
```

Comment:
@monkeyDluffy6017 You're wrong. Sometimes reading code is not enough, and you need to debug it to guarantee your point.

https://github.com/apache/apisix/blob/d8fd3676ede5c07ccce4215f3dfbd823939dcf36/apisix/core/config_etcd.lua#L472

Set up a breakpoint:

```bash
cat > /usr/local/apisix/plugin_inspect_hooks.lua <<EOF
dbg.set_hook("apisix/core/config_etcd.lua", 472, nil, function(info)
    core.log.warn("res: ", tostring(info.vals.res))
    return false
end)
EOF
```

error.log:

```
2023/03/07 16:19:33 [warn] 2472696#2472696: *110 [lua] init.lua:68: setup_hooks(): set hooks: err: true, hooks: ["apisix/core/config_etcd.lua#472"], context: ngx.timer
```

Create a test upstream:

```bash
update_upstream() {
curl http://127.0.0.1:9180/apisix/admin/upstreams/1 \
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "nodes":[
        {
            "host":"httpbin.org",
            "port":80,
            "weight":1
        },
        {
            "host":"postman-echo.com",
            "port":80,
            "weight":1
        }
    ]
}'
}

update_upstream()
```

Override:

```bash
update_upstream()
```

error.log:

```
2023/03/07 16:20:09 [warn] 2472696#2472696: *35 [lua] [string "local dbg = require("apisix.inspect.dbg")..."]:256: res: table: 0x7f5c62fede28, context: ngx.timer
```

Override again:

```bash
update_upstream()
```

error.log:

```
2023/03/07 16:20:25 [warn] 2472696#2472696: *35 [lua] [string "local dbg = require("apisix.inspect.dbg")..."]:256: res: table: 0x7f5c680bbe88, context: ngx.timer
```

Note that the table is different: `old: 0x7f5c62fede28, new: 0x7f5c680bbe88`.

The parent of an upstream is itself,

https://github.com/apache/apisix/blob/d8fd3676ede5c07ccce4215f3dfbd823939dcf36/apisix/upstream.lua#L564

https://github.com/apache/apisix/blob/d8fd3676ede5c07ccce4215f3dfbd823939dcf36/apisix/upstream.lua#L510

So, each time the upstream gets updated, the parent is new, so it's impossible to enter into cases where the parent is the same but with a different checker, except this parallel create_checker bug, and since we do the bugfix, then no need to keep such code anymore. It works for all cases, including route with inline upstream or service with inline upstream.

@spacewander @moonming @membphis I think we should remove such obsolete code to make our code base lint.

Reply:
I see what you mean, except for this concurrent scenario, because the upstream object changes every time, so there can't be a `healthcheck_parent.checker`

Reply:
Yes.

Reply:
Removing this code will result in a lot of ci errors, consider dealing with it in a separate PR

Reply:
@monkeyDluffy6017 @spacewander @moonming @membphis
I take back what I concluded. That code is not dead code, although it's not for concurrent checker creation guard,  it's necessary for service discovery. 
**Each time the discovery result changes, the upstream (inlined in route/service or separate upstream) will do a shallow copy and reserves its parent, so the parent (route/service/upstream) may have a previous checker.**

https://github.com/apache/apisix/blob/c159f04d845c6cdf2dbb32931da7af127bdb60a6/apisix/upstream.lua#L272-L284

Reply:
It looks like that the ci has covered this, so after removed  this code, the ci failed:
https://github.com/apache/apisix/blob/c159f04d845c6cdf2dbb32931da7af127bdb60a6/t/node/healthcheck-discovery.t#L110

Reply:
Yes, so keep that code, although it's only for service discovery but not for normal upstream configuration update from etcd.